File: examples/java/src/main/java/io/superdurable/dex/patterns/parallelsubflows/ExampleSubFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

package io.superdurable.dex.patterns.parallelsubflows;

import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import org.springframework.stereotype.Component;

@Component
public final class ExampleSubFlow implements Flow<String> {
    private final DoWorkStep doWorkStep = new DoWorkStep();

    @Override
    public StepList<String> getSteps() {
        return StepList.startStep(doWorkStep);
    }

    static final class DoWorkStep implements Step<String> {
        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public StepDecision execute(final Context context, final String request) {
            try {
                Thread.sleep(50L + request.length() % 10L * 50L);
            } catch (final InterruptedException interrupted) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("SubFlow work interrupted", interrupted);
            }
            return StepDecision.gracefulComplete(request);
        }
    }
}

File: examples/go/patterns/parallel-subflows/example_subflow.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package parallelsubflows

import (
	"time"

	"github.com/superdurable/dex/sdk-go/dex"
)

type ExampleSubFlow struct{ dex.FlowDefaults }

func NewExampleSubFlow() *ExampleSubFlow { return &ExampleSubFlow{} }

func (*ExampleSubFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{dex.DefineStartStep(doWorkStep{})}
}

func (*ExampleSubFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{}
}

type doWorkStep struct {
	dex.StepDefaultsNoWaitFor[string]
}

func (doWorkStep) GetStepType() string { return "DoWorkStep" }

func (doWorkStep) Execute(_ dex.Context, request string) (*dex.StepDecision, error) {
	time.Sleep(time.Duration(50+len(request)%10*50) * time.Millisecond)
	return dex.GracefulComplete(request), nil
}

var _ dex.Flow = (*ExampleSubFlow)(nil)
